addons/block_code/ui/block_factory.gd

```diff
@@ -0,0 +1,236 @@
+
+	if block_definition == null:
+		push_error("Cannot find a block definition for %s." % block_name)
+		return null
+
+	return instantiate_block(block_definition)
+
+
+func get_block_definition(block_name: String) -> BlockDefinition:
+	var split := block_name.split(":", true, 1)
+
+	if len(split) > 1:
+		return _get_parameter_block_definition(split[0], split[1])
+
+	var block_definition = _get_block_definition(block_name)
+
+	if block_definition == null:
+		# FIXME: This is a workaround for old-style output block references.
+		#        These were generated ahead of time using a block name that has
+		#        a "_" before the parameter name. Now, these parameter blocks
+		#        are generated on demand for any block name containing a ":".
+		#        Please remove this fallback when it is no longer necessary.
+		split = block_name.rsplit("_", true, 1)
+		return _get_parameter_block_definition(split[0], split[1])
+
+	return block_definition
+
+
+func _get_block_definition(block_name: String) -> BlockDefinition:
+	for block_definition in _available_blocks:
+		if block_definition.name == block_name:
+			return block_definition
+	return null
+
+
+func _get_parameter_block_definition(block_name: String, parameter_name: String) -> BlockDefinition:
+	var parent_block_definition := _get_block_definition(block_name)
+
+	if parent_block_definition == null:
+		return null
+
+	# TODO: Get output type by parsing parent block definition's display template like template_editor.gd :/
+
+	var parent_out_parameters = parent_block_definition.get_output_parameters()
+
+	if not parent_out_parameters.has(parameter_name):
+		push_error("The parameter name %s is not an output parameter in %s." % [parameter_name, block_name])
+		return null
+
+	#var parts := reg_match.get_string(1).split(": ")
+	#var param_name := parts[0]
+	var parameter_type: Variant.Type = parent_out_parameters[parameter_name]
+
+	var block_definition := BlockDefinition.new()
+	block_definition.name = &"%s:%s" % [block_name, parameter_name]
+	block_definition.target_node_class = parent_block_definition.target_node_class
+	block_definition.category = parent_block_definition.category
+	block_definition.type = Types.BlockType.VALUE
+	block_definition.variant_type = parameter_type
+	block_definition.display_template = parameter_name
+	block_definition.code_template = parameter_name
+	block_definition.scope = parent_block_definition.code_template
+
+	return block_definition
+
+
+func _init_block_definitions():
+	BlocksCatalog.setup()
+
+	if block_script:
+		var script := _load_object_script()
+		if script and script.has_method("setup_custom_blocks"):
+			script.setup_custom_blocks()
+
+	update_block_definitions()
+
+
+func update_block_definitions():
+	_available_blocks.clear()
+	_available_blocks.append_array(_get_default_block_definitions())
+	_available_blocks.append_array(_get_inherited_block_definitions())
+	_available_blocks.append_array(_get_variable_block_definitions())
```

Comment:
I'll shuffle this around a bit and see if I can flatten these things out a bit and it should reduce the diff a bunch. Mainly this happened because I was frustrated with the number of different places where we gather up block definitions, although that frustration was more with the code before #147. Looking at this at the moment, I'm extremely annoyed that `_get_variable_block_definitions` is accessing `block_script.variables`, so _maaaybe_ BlockScript should be more involved there. I'll leave a note once when I've addressed that!